src/main/BlockPool.cpp

```diff
@@ -0,0 +1,1216 @@
+ *  Copyright (C) 1995, 1996  Robert Gentleman and Ross Ihaka
+ *  Copyright (C) 1998-2007   The R Development Core Team.
+ *  Copyright (C) 2008-2014  Andrew R. Runnalls.
+ *  Copyright (C) 2014 and onwards the Rho Project Authors.
+ *
+ *  Rho is not part of the R project, and bugs and other issues should
+ *  not be reported via r-bugs or other R project channels; instead refer
+ *  to the Rho website.
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+#include "rho/BlockPool.hpp"
+
+#ifdef HAVE_ADDRESS_SANITIZER
+#include "rho/AddressSanitizer.hpp"
+#endif
+
+// The low pointer bits are masked out in the hash function.
+#define LOW_BITS (19)
+#define MAX_COLLISIONS (6)
+#define MAX_REBALANCE_COLLISIONS (3)
+
+// NUM_SMALL_POOLS = (256 / 8) + 1 = 33
+#define NUM_SMALL_POOLS (33)
+
+// Constant indicating a deleted bucket.
+#define EMPTY_KEY (0)
+#define DELETED_KEY (UINTPTR_MAX)
+
+// Samll object arena is 1Gb = 30 bits.
+#define ARENASIZE (1 << 30)
+
+/*
+ * A small object superblock has size 2^18.
+ * Minimum object size = 32 bytes.
+ */
+#define SUPERBLOCK_BITS (18)
+#define SUPERBLOCK_SIZE (1 << SUPERBLOCK_BITS)
+
+/*
+ * Medium object superblocks use 2^19 bytes.
+ * Minimum object size = 64 bytes.
+ */
+#define MEDIUM_SUPERBLOCK_SIZE_LOG2 (19)
+#define MEDIUM_SUPERBLOCK_SIZE (1 << MEDIUM_SUPERBLOCK_SIZE_LOG2)
+
+/*
+ * NOTE: We use a fixed superblock header size, regardless of block size. This
+ * leaves some unused bitset entries for larger block sizes.
+ *
+ * The fixed header size is 1168 bytes. This is the breakdown:
+ *    block_size     =   32 bytes
+ *    next_untouched =   32 bytes
+ *    free_list      =   64 bytes
+ *    free bitset    = 1040 bytes
+ *    total          = 1168 bytes
+ *
+ * The bitset can map 1040 * 8 = 8320 blocks.  The minimum block size for a
+ * 2^18 superblock is 32 bytes, meaning the bitset covers 8320 * 8 = 266240
+ * bytes, which is more than the required range 2^18 - 1168 = 260976.
+ *
+ * For a 2^19 size superblock the minimum object size is 64 bytes,
+ * and 1040 * 8 * 64 = 532480 < 2^19 - 1168. So all objects are mappable
+ * for a 2^19 superblock.
+ */
+#define SUPERBLOCK_HEADER_SIZE (1168)
```

Comment:
Can this be expressed as a calculation, instead of a fixed constant?  Otherwise, it'd need manually updating if we change anything.